### PR TITLE
Don't make tools responsible for checking unknown and renamed lints

### PR DESCRIPTION
**Pull request by jyn514**
Previously, clippy (and any other tool emitting lints) had to have their
own separate UNKNOWN_LINTS pass, because the compiler assumed any tool
lint could be valid. Now, as long as any lint starting with the tool
prefix exists, the compiler will warn when an unknown lint is present.

This may interact with the unstable `tool_lint` feature, which I don't entirely understand, but it will take the burden off those external tools to add their own lint pass, which seems like a step in the right direction to me.

- Don't mark `ineffective_unstable_trait_impl` as an internal lint
- Use clippy's more advanced lint suggestions
- Deprecate the `UNKNOWN_CLIPPY_LINTS` pass (and make it a no-op)
- Say 'unknown lint `clippy::x`' instead of 'unknown lint x'

This is tested by existing clippy tests. When https://github.com/rust-lang/rust/pull/80527 merges, it will also be tested in rustdoc tests. AFAIK there is no way to test this with rustc directly.

### CODE REVIEW

**Comment by flip1995** [comment]
Something that is also important to mention:

`tool_lints` are semi-stable. So you can use them without adding a feature, but only explicitly allowed tools (defined in `is_known_lint_tool` can implement them. This also means, that this is a semi-breaking change. The [reference](https://doc.rust-lang.org/stable/reference/attributes/diagnostics.html#tool-lint-attributes) says:

> If a lint attribute, such as `allow`, references a nonexistent tool lint, the compiler will not warn about the nonexistent lint until you use the tool.

So it doesn't mention, that the handling of missing lints has to be implemented by the tool. With this and the fact, that we know all the tools that implement tool lints, this change should be fine.

**Comment by flip1995** [comment]
So the CI error is because internal lints are only enabled in the `Mode::Rustc`, not when compiling `std`.

This makes the `rustc::existing_doc_keywords` lint useless, since it is only enabled in `std`. But enabling internal lints in `std` results in many warnings / `allow`s of the `rustc::default_hash_types` lint.

**Comment by flip1995** [comment]
Oh well, this uncovered a whole new set of problems, that were silently ignored before: `rustc::internal` lints are also tool lints, but rustc doesn't implement a `unknown_rustc_lints` lint. So writing `deny(rustc::internal)` where internal lints aren't enabled (`std` or even `rustc_macros` for some reason), was just silently ignored. This PR fixes that. But now when a tool like Clippy is compiled with `./x.py`, which also uses internal `rustc` lints it fails, because they are not enabled by tools. They can be enabled by adding `Mode::ToolsRustc` here:

https://github.com/rust-lang/rust/blob/8b002d5c3489a21db2c16e5af63cf5d234f6972c/src/bootstrap/builder.rs#L1268-L1271

But this results in build failures for other tools, that don't adhere to the rustc internal lints, like `rustdoc`.

**Comment by jyn514** [comment]
> But enabling internal lints in std results in many warnings / allows of the rustc::default_hash_types lint.

FWIW, adding `-Wrustc::internal -Arustc::default-hash-types` gives no warnings. So in theory we could special case std like that. I think `existing_doc_keywords` is niche enough it shouldn't be a big deal to remove it though (cc @GuillaumeGomez just in case - I kind of wish you'd pinged me on https://github.com/rust-lang/rust/pull/79541, since I suggested it in https://github.com/rust-lang/rust/pull/79464#issuecomment-734885257).

>  But now when a tool like Clippy is compiled with ./x.py, which also uses internal rustc lints it fails, because they are not enabled by tools. 

I don't quite understand this - I think to test it I would need `x.py build --stage 2`, which takes a while. Can you give an example of a lint that fails?

> But this results in build failures for other tools, that don't adhere to the rustc internal lints, like rustdoc.

I'm happy to fix the lints for rustdoc, I would prefer to have them enabled anyway.

**Comment by GuillaumeGomez** [comment]
`existing_doc_keywords` is aimed to rustc only, not rustdoc or anything else (which means you can still add doc on keywords that don't exist outside of rustc). The goal is to prevent rustc to have doc on keywords which don't exist while allowing other crates to (since it's aimed for proc-macros). So please don't remove it?

**Comment by jyn514** [comment]
@GuillaumeGomez it's aimed at *libstd*, which is the whole point - your PR did nothing because rustc lints aren't applied to libstd currently. I'm ok with denying that lint specifically for std though.

**Comment by GuillaumeGomez** [comment]
libstd and libcore to be exact. If it's denied for both of them then it's all good for me.

**Comment by flip1995** [comment]
> I don't quite understand this - I think to test it I would need `x.py build --stage 2`, which takes a while. Can you give an example of a lint that fails?

So Clippy has `#[deny(rustc::internal)]` in the `lib.rs`. But since for `Mode::ToolsRustc` internal lints aren't enabled currently, the compilation will fail (`./x.py test src/tools/clippy`). Your changes in #80573 should fix this though.

The question then is if other tools want those internal lints.

File: compiler/rustc_lint/src/context.rs [hunk 1]
```diff
@@ -353,10 +353,21 @@ impl LintStore {
...
-        if tool_name.is_some() {
+        if let Some(tool_name) = tool_name {
             match self.by_name.get(&complete_name) {
                 None => match self.lint_groups.get(&*complete_name) {
-                    None => return CheckLintNameResult::Tool(Err((None, String::new()))),
+                    // If the lint isn't registered, there are two possibilities:
+                    None => {
+                        // 1. The tool is currently running, so this lint really doesn't exist.
+                        // FIXME: should this handle tools that never register a lint, like rustfmt?
+                        return if self.by_name.keys().any(|lint| lint.starts_with(&format!("{}::", tool_name))) {
```
Comment (matthewjasper) [inline compiler/rustc_lint/src/context.rs:363]: Could the `format` be moved out of the loop here?

**Review by matthewjasper — COMMENTED** [review]
r=me with nit addressed (and CI passing)

**Comment by jyn514** [comment]
> > I don't quite understand this - I think to test it I would need `x.py build --stage 2`, which takes a while. Can you give an example of a lint that fails?
> 
> So Clippy has `#[deny(rustc::internal)]` in the `lib.rs`. But since for `Mode::ToolsRustc` internal lints aren't enabled currently, the compilation will fail (`./x.py test src/tools/clippy`). Your changes in #80573 should fix this though.
> 
> The question then is if other tools want those internal lints.

So it turns out the issue is that `rustc::ineffective_unstable_trait_impl` is on unconditionally, because it's not actually an internal lint, it's a lint for `staged_api`. So even when internal lints are disabled, my new logic sees `rustc::` and thinks all `rustc::` lints are known.

 FYI @m-ou-se I'm going to change this to just be `ineffective_unstable_trait_impl`, does that sound good? It will still be emitted only for `staged_api`.

There's a second question here which is why all rustc:: lints are *not* known when `-W rustc::internal` isn't passed, but I'm planning to leave that mystery to another day if possible.

**Comment by jyn514** [comment]
https://github.com/rust-lang/rust/pull/76538#issuecomment-690675578:

> Added it to the builtin lints as a rustc:: tool lint, just like the internal lints. That way it doesn't show up in -W help and an #[allow()] attribute for it doesn't trigger a unknown lint error (which happens when bootstrapping rustc+std, using the current beta).

Normally the fix for this is to add `cfg(not(bootstrap))` to the `allow`. I don't have a fix for it showing up in `-W help`, but since people besides the standard library can use staged_api on nightly, I'm ok with adding it there.

File: compiler/rustc_lint/src/context.rs (hunk 1 above)
Comment (jyn514) [reply compiler/rustc_lint/src/context.rs]: Done, thanks.

**Comment by m-ou-se** [comment]
@jyn514 Yeah, that lint is a bit of a weird one. It's mostly/only meant for `std` but still part of an unstable feature usable outside of it. Rustc doesn't really have a concept of "lint that's part of an unstable/private feature". Putting it under `rustc::` was a hack. It seems fine to me to change that.

**Comment by flip1995** [comment]
> lints are _not_ known when `-W rustc::internal` isn't passed, but I'm planning to leave that mystery to another day if possible.

Oh it isn't because of the `-W` flag, it's the `-Zunstable-options` flag, that gates the registration of internal lints. If internal lints were registered unconditionally, it could have a negative impact on performance.

We chose to (ab)use the `unstable-options` option for internal lints, because we didn't think it was worth to add an extra option for it (or smth like this).

**Comment by jyn514** [comment]
Ahh, that makes sense.

So to summary, this does the following things:
1. Move `rustc::ineffective_unstable_trait_impl` to `ineffective_unstable_trait_impl`
2. Move checking of unknown and renamed lints to the lint machinery, so that each new tool doesn't have to add its own pass checking it. Assume that if any `tool::` lint is present, all lints for `tool` are known. In particular this is necessary to avoid warning on `#[deny(rustc::internal)]` when `-Z unstable-options` isn't passed, and in general we have to assume at *some* point that the lints are known or we can't warn about unknown lints.
3. Fix 'unknown lint x' to say 'unknown lint tool::x' instead

@flip1995 does all of that sound good? I think for 2 you'll want to deprecate/remove the clippy pass that currently checks for unknown lints - is that simple to do? Happy to do it myself if you let me know how.

**Comment by flip1995** [comment]
Yes, that sounds good to me.

If everything in this PR works as expected, every error message in the test for ``unknown_clippy_lints`` should be duplicated. If that is the case, I will deprecate the Clippy lint for you (I can either post a patch file or you can give me push access to your fork, so I can push directly to this branch).

File: compiler/rustc_lint/src/context.rs [hunk 2]
```diff
@@ -412,10 +412,15 @@ impl LintStore {
...
-        let symbols = self.by_name.keys().map(|name| Symbol::intern(&name)).collect::<Vec<_>>();
+        let name_lower = lint_name.to_lowercase();
+        let symbols =
+            self.get_lints().iter().map(|l| Symbol::intern(&l.name_lower())).collect::<Vec<_>>();
 
-        let suggestion =
-            find_best_match_for_name(&symbols, Symbol::intern(&lint_name.to_lowercase()), None);
+        if lint_name.chars().any(char::is_uppercase) && self.find_lints(&name_lower).is_ok() {
+            return CheckLintNameResult::NoLint(Some(Symbol::intern(&name_lower)));
+        }
```
Comment (jyn514) [inline compiler/rustc_lint/src/context.rs:421]: This checks if someone mistyped `unused` as `Unused`?

File: src/tools/clippy/clippy_lints/src/deprecated_lints.rs [hunk 3]
```diff
@@ -163,6 +163,19 @@ declare_deprecated_lint! {
...
+    /// **Deprecation reason:** This lint has been uplifted to rustc and is now called
+    /// `panic_fmt`.
     pub PANIC_PARAMS,
     "this lint has been uplifted to rustc and is now called `panic_fmt`"
 }
+
+declare_deprecated_lint! {
+    /// **What it does:** Nothing. This lint has been deprecated.
+    ///
+    /// **Deprecation reason:** This lint has been integrated into the `unknwon_lints`
```
Comment (jyn514) [inline src/tools/clippy/clippy_lints/src/deprecated_lints.rs:177]:
```suggestion
    /// **Deprecation reason:** This lint has been integrated into the `unknown_lints`
```

File: src/tools/clippy/clippy_lints/src/lib.rs [hunk 4]
```diff
@@ -500,6 +500,10 @@ pub fn register_plugins(store: &mut rustc_lint::LintStore, sess: &Session, conf:
         "clippy::panic_params",
         "this lint has been uplifted to rustc and is now called `panic_fmt`",
     );
+    store.register_removed(
+        "clippy::unknown_clippy_lints",
+        "this lint has been integrated into the `unknwon_lints` rustc lint",
```
Comment (jyn514) [inline src/tools/clippy/clippy_lints/src/lib.rs:505]:
```suggestion
        "this lint has been integrated into the `unknown_lints` rustc lint",
```

File: src/tools/clippy/tests/ui/unknown_clippy_lints.stderr [hunk 5]
```diff
@@ -1,52 +1,58 @@
-error: unknown clippy lint: clippy::if_not_els
+error: unknown lint: `clippy::All`
+  --> $DIR/unknown_clippy_lints.rs:5:10
+   |
+LL | #![allow(clippy::All)]
+   |          ^^^^^^^^^^^ help: did you mean: `clippy::all`
+   |
+   = note: `-D unknown-lints` implied by `-D warnings`
```
Comment (jyn514) [inline src/tools/clippy/tests/ui/unknown_clippy_lints.stderr:7]: Hmm, I'm curious to know why this changed. This is a bug fix, right, it should have given a warning before?

Comment (flip1995) [reply src/tools/clippy/tests/ui/unknown_clippy_lints.stderr:7]: Hm? The order is just different. And it has the default weirdness with linting attributes, that it is duplicated for the first inner attribute.

Comment (jyn514) [reply src/tools/clippy/tests/ui/unknown_clippy_lints.stderr:7]: Ah ok, I misread. No need to change anything.

> And it has the default weirdness with linting attributes, that it is duplicated for the first inner attribute.

Does this have an issue open? It seems worth fixing.

Comment (flip1995) [reply src/tools/clippy/tests/ui/unknown_clippy_lints.stderr:7]: I don't know of an issue for this.

Comment (jyn514) [reply src/tools/clippy/tests/ui/unknown_clippy_lints.stderr:7]: I opened https://github.com/rust-lang/rust-clippy/issues/6602.

**Comment by jyn514** [comment]
Alright, I think this is good to go!

@bors r=flip1995,matthewjasper

File: src/tools/clippy/tests/ui/unknown_clippy_lints.stderr (hunk 5 above)
Comment (camelid) [reply src/tools/clippy/tests/ui/unknown_clippy_lints.stderr:7]: I'm not sure if it's just in a different order; at the bottom it shows that it emitted 8 errors previously and now it emits 9.

**Pull request merged**